In [1]:
import pandas as pd

## Question 1

In [2]:
import requests

In [8]:
url = "http://zims-en.kiwix.campusafrica.gos.orange.com/wikipedia_en_all_nopic/A/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [9]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postcode           Borough          Neighbourhood
 0        M1A      Not assigned           Not assigned
 1        M2A      Not assigned           Not assigned
 2        M3A        North York              Parkwoods
 3        M4A        North York       Victoria Village
 4        M5A  Downtown Toronto           Harbourfront
 ..       ...               ...                    ...
 283      M8Z         Etobicoke     The Queensway West
 284      M8Z         Etobicoke  Royal York South West
 285      M8Z         Etobicoke         South of Bloor
 286      M9Z      Not assigned           Not assigned
 287      NaN               NaN                    NaN
 
 [288 rows x 3 columns]]

In [10]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postcode           Borough          Neighbourhood
 0        M1A      Not assigned           Not assigned
 1        M2A      Not assigned           Not assigned
 2        M3A        North York              Parkwoods
 3        M4A        North York       Victoria Village
 4        M5A  Downtown Toronto           Harbourfront
 ..       ...               ...                    ...
 283      M8Z         Etobicoke     The Queensway West
 284      M8Z         Etobicoke  Royal York South West
 285      M8Z         Etobicoke         South of Bloor
 286      M9Z      Not assigned           Not assigned
 287      NaN               NaN                    NaN
 
 [288 rows x 3 columns]]

In [11]:
wiki_data = wiki_data[0]
wiki_data

Postcode           Borough          Neighbourhood
0        M1A      Not assigned           Not assigned
1        M2A      Not assigned           Not assigned
2        M3A        North York              Parkwoods
3        M4A        North York       Victoria Village
4        M5A  Downtown Toronto           Harbourfront
..       ...               ...                    ...
283      M8Z         Etobicoke     The Queensway West
284      M8Z         Etobicoke  Royal York South West
285      M8Z         Etobicoke         South of Bloor
286      M9Z      Not assigned           Not assigned
287      NaN               NaN                    NaN

[288 rows x 3 columns]

In [29]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df = wiki_data[wiki_data["Neighbourhood"] != "Not assigned"]
df = df.dropna()
df

Postcode           Borough             Neighbourhood
2        M3A        North York                 Parkwoods
3        M4A        North York          Victoria Village
4        M5A  Downtown Toronto              Harbourfront
5        M6A        North York          Lawrence Heights
6        M6A        North York            Lawrence Manor
..       ...               ...                       ...
281      M8Z         Etobicoke  Kingsway Park South West
282      M8Z         Etobicoke                 Mimico NW
283      M8Z         Etobicoke        The Queensway West
284      M8Z         Etobicoke     Royal York South West
285      M8Z         Etobicoke            South of Bloor

[209 rows x 3 columns]

In [30]:
df.rename(columns={'Postcode': 'Postal Code'}, inplace=True)
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough             Neighbourhood
2           M3A        North York                 Parkwoods
3           M4A        North York          Victoria Village
4           M5A  Downtown Toronto              Harbourfront
5           M6A        North York          Lawrence Heights
6           M6A        North York            Lawrence Manor
..          ...               ...                       ...
281         M8Z         Etobicoke  Kingsway Park South West
282         M8Z         Etobicoke                 Mimico NW
283         M8Z         Etobicoke        The Queensway West
284         M8Z         Etobicoke     Royal York South West
285         M8Z         Etobicoke            South of Bloor

[201 rows x 3 columns]

In [31]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [32]:
df = df.reset_index()
df

index Postal Code           Borough             Neighbourhood
0        2         M3A        North York                 Parkwoods
1        3         M4A        North York          Victoria Village
2        4         M5A  Downtown Toronto              Harbourfront
3        5         M6A        North York          Lawrence Heights
4        6         M6A        North York            Lawrence Manor
..     ...         ...               ...                       ...
196    281         M8Z         Etobicoke  Kingsway Park South West
197    282         M8Z         Etobicoke                 Mimico NW
198    283         M8Z         Etobicoke        The Queensway West
199    284         M8Z         Etobicoke     Royal York South West
200    285         M8Z         Etobicoke            South of Bloor

[201 rows x 4 columns]

In [33]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough             Neighbourhood
0           M3A        North York                 Parkwoods
1           M4A        North York          Victoria Village
2           M5A  Downtown Toronto              Harbourfront
3           M6A        North York          Lawrence Heights
4           M6A        North York            Lawrence Manor
..          ...               ...                       ...
196         M8Z         Etobicoke  Kingsway Park South West
197         M8Z         Etobicoke                 Mimico NW
198         M8Z         Etobicoke        The Queensway West
199         M8Z         Etobicoke     Royal York South West
200         M8Z         Etobicoke            South of Bloor

[201 rows x 3 columns]

In [34]:
df.shape

(201, 3)

Answer to Question 1: We have 201 rows and 3 columns

### Question 2

In [23]:
pip install geocoder

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
     |████████████████████████████████| 98 kB 7.6 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [24]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [64]:

print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)


The shape of our wiki data is:  (201, 3)
the shape of our csv data is:  (103, 3)


In [36]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [37]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [65]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough             Neighbourhood   Latitude  \
0           M3A        North York                 Parkwoods  43.753259   
1           M4A        North York          Victoria Village  43.725882   
2           M5A  Downtown Toronto              Harbourfront  43.654260   
3           M6A        North York          Lawrence Heights  43.718518   
4           M6A        North York            Lawrence Manor  43.718518   
..          ...               ...                       ...        ...   
196         M8Z         Etobicoke  Kingsway Park South West  43.628841   
197         M8Z         Etobicoke                 Mimico NW  43.628841   
198         M8Z         Etobicoke        The Queensway West  43.628841   
199         M8Z         Etobicoke     Royal York South West  43.628841   
200         M8Z         Etobicoke            South of Bloor  43.628841   

     Longitude  
0   -79.329656  
1   -79.315572  
2   -79.360636  
3   -79.464763  
4   -79.464763  
..         ...  
196 -79.520999  
197 -79.520999  
198 -79.520999  
199 -79.520999  
200 -79.520999  

[201 rows x 5 columns]

In [66]:
combined_data.shape

(201, 5)

Answer to question 2 - We get 201 rows as expected when we do a inner join, so we have good data.

### Question 3

In [67]:
from geopy.geocoders import Nominatim


In [68]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [70]:
!pip install folium

import folium

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
     |████████████████████████████████| 94 kB 3.4 MB/s eta 0:00:011


In [71]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

In [83]:
CLIENT_ID = 'GAP4OI301LO2WXCU5W4WFVFQ5MWJVDIXI0NL55ZH3USG3SFK' 
CLIENT_SECRET = '5EPSD0GTW3YKN3JIDG5LMJJ4PL1UURWNDEB4F4LTF3RVYBWM'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: GAP4OI301LO2WXCU5W4WFVFQ5MWJVDIXI0NL55ZH3USG3SFK
CLIENT_SECRET:5EPSD0GTW3YKN3JIDG5LMJJ4PL1UURWNDEB4F4LTF3RVYBWM


In [84]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']

        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [86]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Harbourfront
Lawrence Heights
Lawrence Manor
Queen's Park
Rouge
Malvern
Don Mills North
Woodbine Gardens
Parkview Hill
Ryerson
Garden District
Glencairn
Cloverdale
Islington
Martin Grove
Princess Gardens
West Deane Park
Highland Creek
Rouge Hill
Port Union
Flemingdon Park
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Bloordale Gardens
Eringate
Markland Wood
Old Burnhamthorpe
Guildwood
Morningside
West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor
Downsview North
Wilson Heights
Thorncliffe Park
Adelaide
King
Richmond
Dovercourt Village
Dufferin
Scarborough Village
Fairview
Henry Farm
Oriole
Northwood Park
York University
East Toronto
Harbourfront East
Toronto Islands
Union Station
Little Portugal
Trinity
East Birchmount Park
Ionview
Kennedy Park
Bayview Village
CFB Toronto
Downsview East
The Danforth West
Riverdale
Design Exchange
Toronto Dominion Cent

In [87]:
venues_in_toronto.shape

(2405, 5)

In [88]:

venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0                     KFC   Fast Food Restaurant  
1         Brookbanks Park                   Park  
2           Variety Store      Food & Drink Shop  
3  Victoria Village Arena           Hockey Arena  
4               Portugril  Portuguese Restaurant

In [89]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0            Parkwoods               43.753259               -79.329656   
1            Parkwoods               43.753259               -79.329656   
2            Parkwoods               43.753259               -79.329656   
3     Victoria Village               43.725882               -79.315572   
4     Victoria Village               43.725882               -79.315572   
...                ...                     ...                      ...   
2393    South of Bloor               43.628841               -79.520999   
2394    South of Bloor               43.628841               -79.520999   
2395    South of Bloor               43.628841               -79.520999   
2396    South of Bloor               43.628841               -79.520999   
2397    South of Bloor               43.628841               -79.520999   

                       Venue         Venue Category  
0                        KFC   Fast Food Restaurant  
1            Brookbanks Park                   Park  
2              Variety Store      Food & Drink Shop  
3     Victoria Village Arena           Hockey Arena  
4                  Portugril  Portuguese Restaurant  
...                      ...                    ...  
2393        South St. Burger           Burger Joint  
2394              Wingporium            Wings Joint  
2395          Healthy Planet        Supplement Shop  
2396               Dollarama         Discount Store  
2397                  Subway         Sandwich Place  

[821 rows x 5 columns]

In [90]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                                     
Accessories Store                Wexford               43.750072   
Adult Boutique      Church and Wellesley               43.665860   
Airport                 King and Spadina               43.737473   
Airport Food Court      King and Spadina               43.628947   
Airport Gate            King and Spadina               43.628947   
...                                  ...                     ...   
Wine Bar                  Victoria Hotel               43.653206   
Wine Shop           Church and Wellesley               43.665860   
Wings Joint           The Queensway West               43.628841   
Women's Store        Caledonia-Fairbanks               43.689026   
Yoga Studio                      Trinity               43.705369   

                    Neighbourhood Longitude                            Venue  
Venue Category                                                                
Accessories Store                -79.295849                      Puffin Gear  
Adult Boutique                   -79.383160                        Seduction  
Airport                          -79.394420  Toronto Downsview Airport (YZD)  
Airport Food Court               -79.394420                Billy Bishop Café  
Airport Gate                     -79.394420                           Gate 8  
...                                     ...                              ...  
Wine Bar                         -79.379817                The National Club  
Wine Shop                        -79.383160                        Wine Rack  
Wings Joint                      -79.520999                       Wingporium  
Women's Store                    -79.453512                    Maximum Woman  
Yoga Studio                      -79.321558                        YogaSpace  

[224 rows x 4 columns]

In [91]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
2400                  0               0        0                   0   
2401                  0               0        0                   0   
2402                  0               0        0                   0   
2403                  0               0        0                   0   
2404                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
2400             0               0                0                 0   
2401             0               0                0                 0   
2402             0               0                0                 0   
2403             0               0                0                 0   
2404             0               0                0                 0   

      American Restaurant  Aquarium  ...  Vegetarian / Vegan Restaurant  \
0                       0         0  ...                              0   
1                       0         0  ...                              0   
2                       0         0  ...                              0   
3                       0         0  ...                              0   
4                       0         0  ...                              0   
...                   ...       ...  ...                            ...   
2400                    0         0  ...                              0   
2401                    0         0  ...                              0   
2402                    0         0  ...                              0   
2403                    0         0  ...                              0   
2404                    0         0  ...                              0   

      Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                    0            0                      0                0   
1                    0            0                      0                0   
2                    0            0                      0                0   
3                    0            0                      0                0   
4                    0            0                      0                0   
...                ...          ...                    ...              ...   
2400                 0            0                      0                0   
2401                 0            0                      0                0   
2402                 0            0                      0                0   
2403                 0            0                      0                0   
2404                 0            0                      0                0   

      Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0            0          0            0              0            0  
1            0          0            0              0            0  
2            0          0            0              0            0  
3            0          0            0              0            0  
4            0          0            0              0            0  


In [99]:

toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Yoga Studio     Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0            0         Parkwoods                  0               0        0   
1            0         Parkwoods                  0               0        0   
2            0         Parkwoods                  0               0        0   
3            0  Victoria Village                  0               0        0   
4            0  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  ...  Truck Stop  Vegetarian / Vegan Restaurant  \
0                 0  ...           0                              0   
1                 0  ...           0                              0   
2                 0  ...           0                              0   
3                 0  ...           0                              0   
4                 0  ...           0                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  
0         0          0            0              0  
1         0          0            0              0  
2         0          0            0              0  
3         0          0            0              0  
4         0          0            0              0  

[5 rows x 225 columns]

In [100]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Yoga Studio  Accessories Store  Adult Boutique  Airport  \
0         Adelaide          0.0                0.0             0.0      0.0   
1        Agincourt          0.0                0.0             0.0      0.0   
2  Agincourt North          0.0                0.0             0.0      0.0   
3   Albion Gardens          0.0                0.0             0.0      0.0   
4        Alderwood          0.0                0.0             0.0      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  ...  Truck Stop  Vegetarian / Vegan Restaurant  \
0               0.0  ...         0.0                       0.033333   
1               0.0  ...         0.0                       0.000000   
2               0.0  ...         0.0                       0.000000   
3               0.0  ...         0.0                       0.000000   
4               0.0  ...         0.0                       0.000000   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0               0.0          0.0                    0.0              0.0   
1               0.0          0.0                    0.0              0.0   
2               0.0          0.0                    0.0              0.0   
3               0.0          0.0                    0.0              0.0   
4               0.0          0.0                    0.0              0.0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  
0       0.0        0.0          0.0            0.0  
1       0.0        0.0          0.0            0.0  
2       0.0        0.0          0.0            0.0  
3       0.0        0.0          0.0            0.0  
4       0.0        0.0          0.0            0.0  

[5 rows x 225 columns]

In [101]:

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [102]:
import numpy as np

In [103]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0         Adelaide                  Café           Coffee Shop   
1        Agincourt                Lounge        Clothing Store   
2  Agincourt North            Playground                  Park   
3   Albion Gardens         Grocery Store  Fast Food Restaurant   
4        Alderwood           Pizza Place                   Pub   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Sushi Restaurant   Japanese Restaurant            Restaurant   
1  Latin American Restaurant          Skating Rink        Breakfast Spot   
2               Intersection         Women's Store      Department Store   
3                Pizza Place           Coffee Shop        Sandwich Place   
4               Skating Rink        Sandwich Place          Dance Studio   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0            Steakhouse                Bakery                    Speakeasy   
1         Women's Store                 Diner                  Escape Room   
2           Escape Room     Electronics Store  Eastern European Restaurant   
3            Beer Store              Pharmacy                    Gastropub   
4           Coffee Shop            Playground              College Stadium   

         9th Most Common Venue       10th Most Common Venue  
0                    Juice Bar           Seafood Restaurant  
1            Electronics Store  Eastern European Restaurant  
2                    Drugstore                   Donut Shop  
3                  Gas Station                   Donut Shop  
4  Eastern European Restaurant                    Drugstore

In [105]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [106]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [107]:
kmeans.labels_[0:100]

array([2, 2, 1, 3, 3, 3, 2, 2, 3, 3, 2, 2, 3, 2, 2, 1, 2, 2, 1, 3, 2, 3,
       2, 2, 2, 2, 3, 2, 2, 2, 4, 2, 2, 2, 3, 3, 2, 2, 2, 2, 2, 1, 2, 1,
       3, 2, 1, 2, 3, 1, 0, 3, 2, 3, 2, 2, 2, 3, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 3, 2, 2, 3, 0, 3, 2, 3, 0, 2, 3, 3, 2, 4, 3, 3, 3, 2, 2, 2,
       0, 1, 0, 2, 1, 3, 3, 3, 3, 1, 3, 2], dtype=int32)

In [108]:

neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [109]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough     Neighbourhood   Latitude  Longitude  \
0         M3A        North York         Parkwoods  43.753259 -79.329656   
1         M4A        North York  Victoria Village  43.725882 -79.315572   
2         M5A  Downtown Toronto      Harbourfront  43.654260 -79.360636   
3         M6A        North York  Lawrence Heights  43.718518 -79.464763   
4         M6A        North York    Lawrence Manor  43.718518 -79.464763   

   Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
0             1.0  Fast Food Restaurant                   Park   
1             3.0         Grocery Store  Portuguese Restaurant   
2             2.0           Coffee Shop                 Bakery   
3             3.0        Clothing Store  Vietnamese Restaurant   
4             3.0        Clothing Store  Vietnamese Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Food & Drink Shop         Women's Store          Dessert Shop   
1          Hockey Arena            Nail Salon           Coffee Shop   
2                  Park                   Pub                   Spa   
3    Miscellaneous Shop              Boutique           Coffee Shop   
4    Miscellaneous Shop              Boutique           Coffee Shop   

    6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0             Escape Room     Electronics Store  Eastern European Restaurant   
1            Concert Hall          Dessert Shop                  Escape Room   
2          Farmers Market                  Bank                         Café   
3  Furniture / Home Store     Accessories Store            Electronics Store   
4  Furniture / Home Store     Accessories Store            Electronics Store   

  9th Most Common Venue       10th Most Common Venue  
0             Drugstore                   Donut Shop  
1     Electronics Store  Eastern European Restaurant  
2        Breakfast Spot            French Restaurant  
3           Escape Room           Dim Sum Restaurant  
4           Escape Room           Dim Sum Restaurant

In [110]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [111]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [112]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Cluster 1

In [115]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
104  North York             0.0        Baseball Field          Food Service   
105  North York             0.0        Baseball Field          Food Service   
191   Etobicoke             0.0        Baseball Field         Women's Store   
192   Etobicoke             0.0        Baseball Field         Women's Store   
193   Etobicoke             0.0        Baseball Field         Women's Store   
194   Etobicoke             0.0        Baseball Field         Women's Store   
195   Etobicoke             0.0        Baseball Field         Women's Store   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
104         Women's Store                 Diner           Event Space   
105         Women's Store                 Diner           Event Space   
191        Farmers Market           Event Space  Ethiopian Restaurant   
192        Farmers Market           Event Space  Ethiopian Restaurant   
193        Farmers Market           Event Space  Ethiopian Restaurant   
194        Farmers Market           Event Space  Ethiopian Restaurant   
195        Farmers Market           Event Space  Ethiopian Restaurant   

    6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
104  Ethiopian Restaurant           Escape Room            Electronics Store   
105  Ethiopian Restaurant           Escape Room            Electronics Store   
191           Escape Room     Electronics Store  Eastern European Restaurant   
192           Escape Room     Electronics Store  Eastern European Restaurant   
193           Escape Room     Electronics Store  Eastern European Restaurant   
194           Escape Room     Electronics Store  Eastern European Restaurant   
195           Escape Room     Electronics Store  Eastern European Restaurant   

           9th Most Common Venue 10th Most Common Venue  
104  Eastern European Restaurant              Drugstore  
105  Eastern European Restaurant              Drugstore  
191                    Drugstore             Donut Shop  
192                    Drugstore             Donut Shop  
193                    Drugstore             Donut Shop  
194                    Drugstore             Donut Shop  
195                    Drugstore             Donut Shop

In [114]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0          North York             1.0  Fast Food Restaurant   
36               York             1.0                  Park   
58          East York             1.0                  Park   
68         North York             1.0                  Park   
69         North York             1.0                  Park   
82         North York             1.0         Grocery Store   
87         North York             1.0                  Park   
88         North York             1.0                  Park   
89         North York             1.0                  Park   
94         North York             1.0                  Park   
95         North York             1.0                  Park   
110   Central Toronto             1.0      Business Service   
114              York             1.0                  Park   
118        North York             1.0                  Park   
135         Etobicoke             1.0                  Park   
136         Etobicoke             1.0                  Park   
137         Etobicoke             1.0                  Park   
138         Etobicoke             1.0                  Park   
148   Central Toronto             1.0          Tennis Court   
149   Central Toronto             1.0          Tennis Court   
153       Scarborough             1.0            Playground   
154       Scarborough             1.0            Playground   
155       Scarborough             1.0            Playground   
156       Scarborough             1.0            Playground   
176  Downtown Toronto             1.0                  Park   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                          Park     Food & Drink Shop         Women's Store   
36                Women's Store                  Pool      Department Store   
58            Convenience Store         Women's Store          Dessert Shop   
68                      Airport     Other Repair Shop      Business Service   
69                      Airport     Other Repair Shop      Business Service   
82                         Park                  Bank         Shopping Mall   
87   Construction & Landscaping      Basketball Court                Bakery   
88   Construction & Landscaping      Basketball Court                Bakery   
89   Construction & Landscaping      Basketball Court                Bakery   
94                Women's Store          Dessert Shop  Ethiopian Restaurant   
95                Women's Store          Dessert Shop  Ethiopian Restaurant   
110                        Park           Swim School              Bus Line   
114               Women's Store          Dessert Shop  Ethiopian Restaurant   
118           Convenience Store         Women's Store          Dessert Shop   
135           Mobile Phone Shop        Sandwich Place         Women's Store   
136           Mobile Phone Shop        Sandwich Place         Women's Store   
137           Mobile Phone Shop        Sandwich Place         Women's Store   
138           Mobile Phone Shop        Sandwich Place         Women's Store   
148                        Park            Playground                 Trail   
149                        Park            Playground                 Trail   
153                        Park          Intersection         Women's Store   
154                        Park          Intersection         Women's Store   
155                        Park          Intersection         Women's Store   
156                        Park          Intersection         Women's Store   
176                  Playground                 Trail         Women's Store   

    5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0            Dessert Shop           Escape Room            Electronics Store   
36            Escape Room     Electronics Store  Eastern European Restaurant   
58   Ethiopian Restaurant           Escape Room            Electronics Store   


Cluster 2

In [119]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0          North York             1.0  Fast Food Restaurant   
36               York             1.0                  Park   
58          East York             1.0                  Park   
68         North York             1.0                  Park   
69         North York             1.0                  Park   
82         North York             1.0         Grocery Store   
87         North York             1.0                  Park   
88         North York             1.0                  Park   
89         North York             1.0                  Park   
94         North York             1.0                  Park   
95         North York             1.0                  Park   
110   Central Toronto             1.0      Business Service   
114              York             1.0                  Park   
118        North York             1.0                  Park   
135         Etobicoke             1.0                  Park   
136         Etobicoke             1.0                  Park   
137         Etobicoke             1.0                  Park   
138         Etobicoke             1.0                  Park   
148   Central Toronto             1.0          Tennis Court   
149   Central Toronto             1.0          Tennis Court   
153       Scarborough             1.0            Playground   
154       Scarborough             1.0            Playground   
155       Scarborough             1.0            Playground   
156       Scarborough             1.0            Playground   
176  Downtown Toronto             1.0                  Park   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                          Park     Food & Drink Shop         Women's Store   
36                Women's Store                  Pool      Department Store   
58            Convenience Store         Women's Store          Dessert Shop   
68                      Airport     Other Repair Shop      Business Service   
69                      Airport     Other Repair Shop      Business Service   
82                         Park                  Bank         Shopping Mall   
87   Construction & Landscaping      Basketball Court                Bakery   
88   Construction & Landscaping      Basketball Court                Bakery   
89   Construction & Landscaping      Basketball Court                Bakery   
94                Women's Store          Dessert Shop  Ethiopian Restaurant   
95                Women's Store          Dessert Shop  Ethiopian Restaurant   
110                        Park           Swim School              Bus Line   
114               Women's Store          Dessert Shop  Ethiopian Restaurant   
118           Convenience Store         Women's Store          Dessert Shop   
135           Mobile Phone Shop        Sandwich Place         Women's Store   
136           Mobile Phone Shop        Sandwich Place         Women's Store   
137           Mobile Phone Shop        Sandwich Place         Women's Store   
138           Mobile Phone Shop        Sandwich Place         Women's Store   
148                        Park            Playground                 Trail   
149                        Park            Playground                 Trail   
153                        Park          Intersection         Women's Store   
154                        Park          Intersection         Women's Store   
155                        Park          Intersection         Women's Store   
156                        Park          Intersection         Women's Store   
176                  Playground                 Trail         Women's Store   

    5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0            Dessert Shop           Escape Room            Electronics Store   
36            Escape Room     Electronics Store  Eastern European Restaurant   
58   Ethiopian Restaurant           Escape Room            Electronics Store   


Cluster 3

In [117]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2    Downtown Toronto             2.0           Coffee Shop   
8          North York             2.0                   Gym   
11   Downtown Toronto             2.0                  Café   
12   Downtown Toronto             2.0                  Café   
13         North York             2.0           Pizza Place   
..                ...             ...                   ...   
196         Etobicoke             2.0         Grocery Store   
197         Etobicoke             2.0         Grocery Store   
198         Etobicoke             2.0         Grocery Store   
199         Etobicoke             2.0         Grocery Store   
200         Etobicoke             2.0         Grocery Store   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                  Bakery                  Park                   Pub   
8     Japanese Restaurant        Baseball Field  Caribbean Restaurant   
11            Coffee Shop        Clothing Store   Japanese Restaurant   
12            Coffee Shop        Clothing Store   Japanese Restaurant   
13     Italian Restaurant            Playground                Bakery   
..                    ...                   ...                   ...   
196         Burrito Place         Tanning Salon          Burger Joint   
197         Burrito Place         Tanning Salon          Burger Joint   
198         Burrito Place         Tanning Salon          Burger Joint   
199         Burrito Place         Tanning Salon          Burger Joint   
200         Burrito Place         Tanning Salon          Burger Joint   

         5th Most Common Venue 6th Most Common Venue  \
2                          Spa        Farmers Market   
8                         Café  Ethiopian Restaurant   
11   Middle Eastern Restaurant               Theater   
12   Middle Eastern Restaurant               Theater   
13                 Escape Room     Electronics Store   
..                         ...                   ...   
196             Sandwich Place                Bakery   
197             Sandwich Place                Bakery   
198             Sandwich Place                Bakery   
199             Sandwich Place                Bakery   
200             Sandwich Place                Bakery   

           7th Most Common Venue 8th Most Common Venue  \
2                           Bank                  Café   
8                    Escape Room     Electronics Store   
11                    Comic Shop       Thai Restaurant   
12                    Comic Shop       Thai Restaurant   
13   Eastern European Restaurant             Drugstore   
..                           ...                   ...   
196              Supplement Shop           Flower Shop   
197              Supplement Shop           Flower Shop   
198              Supplement Shop           Flower Shop   
199              Supplement Shop           Flower Shop   
200              Supplement Shop           Flower Shop   

           9th Most Common Venue 10th Most Common Venue  
2                 Breakfast Spot      French Restaurant  
8    Eastern European Restaurant              Drugstore  
11                Sandwich Place           Burger Joint  
12                Sandwich Place           Burger Joint  
13                    Donut Shop                Dog Run  
..                           ...                    ...  
196               Discount Store      Convenience Store  
197               Discount Store      Convenience Store  
198               Discount Store      Convenience Store  
199               Discount Store      Convenience Store  
200               Discount Store      Convenience Store  

[95 rows x 12 columns]

Cluster 4

In [118]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
1     North York             3.0         Grocery Store  Portuguese Restaurant   
3     North York             3.0        Clothing Store  Vietnamese Restaurant   
4     North York             3.0        Clothing Store  Vietnamese Restaurant   
6    Scarborough             3.0  Fast Food Restaurant     Falafel Restaurant   
7    Scarborough             3.0  Fast Food Restaurant     Falafel Restaurant   
..           ...             ...                   ...                    ...   
173    Etobicoke             3.0         Grocery Store   Fast Food Restaurant   
174    Etobicoke             3.0         Grocery Store   Fast Food Restaurant   
175  Scarborough             3.0  Fast Food Restaurant            Pizza Place   
178    Etobicoke             3.0           Pizza Place                    Pub   
179    Etobicoke             3.0           Pizza Place                    Pub   

    3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
1            Hockey Arena              Nail Salon           Coffee Shop   
3      Miscellaneous Shop                Boutique           Coffee Shop   
4      Miscellaneous Shop                Boutique           Coffee Shop   
6             Event Space    Ethiopian Restaurant           Escape Room   
7             Event Space    Ethiopian Restaurant           Escape Room   
..                    ...                     ...                   ...   
173           Pizza Place             Coffee Shop        Sandwich Place   
174           Pizza Place             Coffee Shop        Sandwich Place   
175                  Bank  Furniture / Home Store            Nail Salon   
178          Skating Rink          Sandwich Place          Dance Studio   
179          Skating Rink          Sandwich Place          Dance Studio   

      6th Most Common Venue        7th Most Common Venue  \
1              Concert Hall                 Dessert Shop   
3    Furniture / Home Store            Accessories Store   
4    Furniture / Home Store            Accessories Store   
6         Electronics Store  Eastern European Restaurant   
7         Electronics Store  Eastern European Restaurant   
..                      ...                          ...   
173              Beer Store                     Pharmacy   
174              Beer Store                     Pharmacy   
175            Noodle House                     Pharmacy   
178             Coffee Shop                   Playground   
179             Coffee Shop                   Playground   

    8th Most Common Venue        9th Most Common Venue  \
1             Escape Room            Electronics Store   
3       Electronics Store                  Escape Room   
4       Electronics Store                  Escape Room   
6               Drugstore                   Donut Shop   
7               Drugstore                   Donut Shop   
..                    ...                          ...   
173             Gastropub                  Gas Station   
174             Gastropub                  Gas Station   
175           Coffee Shop           Chinese Restaurant   
178       College Stadium  Eastern European Restaurant   
179       College Stadium  Eastern European Restaurant   

          10th Most Common Venue  
1    Eastern European Restaurant  
3             Dim Sum Restaurant  
4             Dim Sum Restaurant  
6                        Dog Run  
7                        Dog Run  
..                           ...  
173                   Donut Shop  
174                   Donut Shop  
175               Sandwich Place  
178                    Drugstore  
179                    Drugstore  

[65 rows x 12 columns]